### ISSUE TITLE
java.lang.IndexOutOfBoundsException with v0.6.2

### ISSUE THREAD

**Issue by shreknitin89** [jenkinsci/ansicolor-plugin#158]
I used to get similar exception with previous versions but that got fixed in previous versions. Since the latest version I am getting the below exception while opening in the middle of build running. And in some cases after the build is finished the tail end of the logs were cut off. Below is the stacktrace 

`java.lang.IndexOutOfBoundsException
	at hudson.MarkupText.addMarkup(MarkupText.java:260)
	at hudson.plugins.ansicolor.ColorConsoleAnnotator$1EmitterImpl.emitHtml(ColorConsoleAnnotator.java:100)
	at hudson.plugins.ansicolor.AnsiAttributeElement.emitOpen(AnsiAttributeElement.java:47)
	at hudson.plugins.ansicolor.AnsiHtmlOutputStream.openTag(AnsiHtmlOutputStream.java:119)
	at hudson.plugins.ansicolor.AnsiHtmlOutputStream.processSetAttribute(AnsiHtmlOutputStream.java:428)
	at hudson.plugins.ansicolor.AnsiOutputStream.processEscapeCommand(AnsiOutputStream.java:361)
	at hudson.plugins.ansicolor.AnsiOutputStream.write(AnsiOutputStream.java:138)
	at hudson.plugins.ansicolor.AnsiHtmlOutputStream.write(AnsiHtmlOutputStream.java:254)
	at org.apache.commons.io.output.ProxyOutputStream.write(ProxyOutputStream.java:54)
	at hudson.plugins.ansicolor.ColorConsoleAnnotator.annotate(ColorConsoleAnnotator.java:133)
	at hudson.console.ConsoleAnnotator$ConsoleAnnotatorAggregator.annotate(ConsoleAnnotator.java:108)
	at hudson.console.ConsoleAnnotationOutputStream.eol(ConsoleAnnotationOutputStream.java:146)
	at hudson.console.LineTransformationOutputStream.eol(LineTransformationOutputStream.java:60)
	at hudson.console.LineTransformationOutputStream.write(LineTransformationOutputStream.java:56)
	at hudson.console.LineTransformationOutputStream.write(LineTransformationOutputStream.java:74)
	at org.apache.commons.io.output.ProxyOutputStream.write(ProxyOutputStream.java:89)
	at org.kohsuke.stapler.framework.io.LargeText$HeadMark.moveTo(LargeText.java:314)
	at org.kohsuke.stapler.framework.io.LargeText.writeLogTo(LargeText.java:229)
	at hudson.console.AnnotatedLargeText.writeHtmlTo(AnnotatedLargeText.java:175)
	at hudson.console.AnnotatedLargeText.writeLogTo(AnnotatedLargeText.java:147)
	at org.kohsuke.stapler.framework.io.LargeText.doProgressText(LargeText.java:264)
	at hudson.console.AnnotatedLargeText.doProgressiveHtml(AnnotatedLargeText.java:95)
	at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627)
	at org.kohsuke.stapler.Function$MethodFunction.invoke(Function.java:396)
	at org.kohsuke.stapler.Function$InstanceFunction.invoke(Function.java:408)
	at org.kohsuke.stapler.Function.bindAndInvoke(Function.java:212)
	at org.kohsuke.stapler.Function.bindAndInvokeAndServeResponse(Function.java:145)
	at org.kohsuke.stapler.MetaClass$11.doDispatch(MetaClass.java:535)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:747)
Caused: javax.servlet.ServletException
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:797)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:878)
	at org.kohsuke.stapler.MetaClass$2.doDispatch(MetaClass.java:219)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:747)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:878)
	at org.kohsuke.stapler.MetaClass$9.dispatch(MetaClass.java:456)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:747)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:878)
	at org.kohsuke.stapler.MetaClass$4.doDispatch(MetaClass.java:280)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:747)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:878)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:676)
	at org.kohsuke.stapler.Stapler.service(Stapler.java:238)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:873)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1623)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:154)
	at jenkins.telemetry.impl.UserLanguages$AcceptLanguageFilter.doFilter(UserLanguages.java:128)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at hudson.util.PluginServletFilter.doFilter(PluginServletFilter.java:157)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
	at hudson.security.csrf.CrumbFilter.doFilter(CrumbFilter.java:99)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:84)
	at hudson.security.UnwrapSecurityExceptionFilter.doFilter(UnwrapSecurityExceptionFilter.java:51)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:117)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.providers.anonymous.AnonymousProcessingFilter.doFilter(AnonymousProcessingFilter.java:125)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.rememberme.RememberMeProcessingFilter.doFilter(RememberMeProcessingFilter.java:142)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.AbstractProcessingFilter.doFilter(AbstractProcessingFilter.java:271)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.BasicHeaderProcessor.doFilter(BasicHeaderProcessor.java:93)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.context.HttpSessionContextIntegrationFilter.doFilter(HttpSessionContextIntegrationFilter.java:249)
	at hudson.security.HttpSessionContextIntegrationFilter2.doFilter(HttpSessionContextIntegrationFilter2.java:67)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at hudson.security.ChainedServletFilter.doFilter(ChainedServletFilter.java:90)
	at hudson.security.HudsonFilter.doFilter(HudsonFilter.java:171)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
	at org.kohsuke.stapler.compression.CompressionFilter.doFilter(CompressionFilter.java:49)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
	at hudson.util.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:82)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
	at org.kohsuke.stapler.DiagnosticThreadNameFilter.doFilter(DiagnosticThreadNameFilter.java:30)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:540)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:146)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:524)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:257)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1700)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1345)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:480)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1667)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1247)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:505)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:370)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:267)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:698)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:804)
	at java.lang.Thread.run(Thread.java:745)`

### Test environment

- Jenkins Version: 2.179
- AnsiColor Version: 0.6.2

The setup was simple

<img width="1439" alt="Screen Shot 2019-05-30 at 4 04 42 PM" src="https://user-images.githubusercontent.com/16711885/58670658-b44bac00-82f4-11e9-9eb5-abc0b96ecbcb.png">

**Comment by shreknitin89** [comment]
@dwnusbaum Any update on this issue?

**Comment by dwnusbaum** [comment]
@shreknitin89 No, I am not working on it and do not have any plans to do so. If you can create a failing test case in [AnsiColorBuildWrapperTest.java](https://github.com/jenkinsci/ansicolor-plugin/blob/ff2264aa5d96ab819f96b904fec1766d529d194b/src/test/java/hudson/plugins/ansicolor/AnsiColorBuildWrapperTest.java) it is much more likely that someone would be able to spend some of their free time to fix the issue since they would be able to easily reproduce the issue from scratch. The test case for the similar issue from #139 is [AnsiColorBuildWrapperTest.testNonAscii](https://github.com/jenkinsci/ansicolor-plugin/blob/ff2264aa5d96ab819f96b904fec1766d529d194b/src/test/java/hudson/plugins/ansicolor/AnsiColorBuildWrapperTest.java#L175-L201), perhaps it would be useful when constructing a test case.

If I had to guess, there is some kind of unusual text in your logs that is causing this problem. If you see some log output before the exception when it happens, then it would be useful to get those lines and the ones immediately afterward from the log file on disk and post them here so they could be used as the basis of a test case.

**Comment by tszmytka** [comment]
@shreknitin89 Could you please provide the actual log from the job which suffers from this exception? It is impossible to provide a fix for something we cannot reproduce. Simply setting up a build job with the Jenkins and ansi-color plugin versions you provided doesn't render the exception.

**Comment by fcristovao** [comment]
Hey there.
I've also hit this problem.

Test environment
Jenkins Version: 2.204.5
AnsiColor Version: 0.6.3

Here's the stacktrace:
```
java.lang.IndexOutOfBoundsException
	at hudson.MarkupText.rangeCheck(MarkupText.java:276)
	at hudson.MarkupText.addMarkup(MarkupText.java:258)
	at hudson.plugins.ansicolor.ColorConsoleAnnotator$1EmitterImpl.hideIfNeeded(ColorConsoleAnnotator.java:119)
	at hudson.plugins.ansicolor.ColorConsoleAnnotator$1EmitterImpl.emitInvisibleSequence(ColorConsoleAnnotator.java:127)
	at hudson.plugins.ansicolor.AnsiHtmlOutputStream.closeOpenTags(AnsiHtmlOutputStream.java:127)
	at hudson.plugins.ansicolor.AnsiHtmlOutputStream.processAttributeRest(AnsiHtmlOutputStream.java:576)
	at hudson.plugins.ansicolor.AnsiOutputStream.processEscapeCommand(AnsiOutputStream.java:358)
	at hudson.plugins.ansicolor.AnsiOutputStream.write(AnsiOutputStream.java:138)
	at hudson.plugins.ansicolor.AnsiHtmlOutputStream.write(AnsiHtmlOutputStream.java:254)
	at org.apache.commons.io.output.ProxyOutputStream.write(ProxyOutputStream.java:54)
```

And I'm attaching the minified output that made it break (I think): [scratch_42.txt](https://github.com/jenkinsci/ansicolor-plugin/files/4506081/scratch_42.txt)

Once the build is finished, the output is actually displayed correctly (up to the offending line)

**Comment by tszmytka** [comment]
@fcristovao Thanks for a comprehensive report.
You actually found a separate (although similar) issue. An `IndexOutOfBoundsException` is thrown in 2 methods:
`hudson.MarkupText.addMarkup` 
`hudson.MarkupText.rangeCheck` 
The original reporter managed to trigger the first instance and you - the second one.

More importantly, the output you attached allows to reproduce both (the original one with slight modification) so I've just added 2 test cases.
A fix will follow.

**Comment by tszmytka** [comment]
Fixed and released in [v0.7.0](https://github.com/jenkinsci/ansicolor-plugin/releases/tag/ansicolor-0.7.0)
